Friendly Revision Labels

manfromjupyter opened the pull request:
Done:
- Ability to set, edit, and delete friendly labels on revision page for each label.
- Ability to set a friendly label (showing you the current one) on Upload New Version and on Approve Change Request forms.
- Made placeholder text consistent for optional fields
- Comparing versions uses the friendly label where it exists. Same for the banners and everything else.
- Made readonly fields consistent in code.
- Started a refactor on how we link things (to do things a way all frameworks accept and allow us more control, not just doing it the Rails way).
- Users can reset the version back to original UUID by setting an empty label. 
- Successfully sets all friendly label saves as inline alerts that doesn't reload page (to clean up one's versions super quickly).
- MAKE SURE that users still always have a way to still see the full original UUID.
- Fixed issue where revisions page with no comparisons to show was breaking datatable to fire a JS error (not rendering search for the table, not rendering pagination, and not added various classes to the table elements).
- Added error catch if trying to set version to one that already exists or setting one longer than 12 characters.
- Gave approver of change requests the decision of what version label to set it to (not the requestor)

Will do in another ticket:
- (To make consistent with label behavior) Saving/editing **revision summaries** send inline-alerts without it reloading page (so users can clean up summaries of all change logs of their notebooks super quickly).
- (To un-Ruby how we do links so it's more transferrable to other frameworks) Change how we do our `link_to_notebooks`, `link_to_groups`, and `link_to_users` to instead use the `_link.slim` way. This way also works in frameworks like Django Python so see it as a less-Ruby-only way. Will also give us more control over these links (HTML-wise, style-wise, and accessibility-wise).

Closes: #181

mcrutch commented:
Reviews table does not respect custom version numbers 
* notebooks/#####/reviews
* /reviews

mcrutch commented:
Do not do this:
> (To un-Ruby how we do links so it's more transferrable to other frameworks) Change how we do our link_to_notebooks, link_to_groups, and link_to_users to instead use the _link.slim way. This way also works in frameworks like Django Python so see it as a less-Ruby-only way. Will also give us more control over these links (HTML-wise, style-wise, and accessibility-wise).

Without ensuring you factor in what would be required for the #718 merge request.  Those magic helpers exist for a reason.

mcrutch commented on app/controllers/application_controller.rb:
```diff
@@ -580,6 +580,26 @@ def verify_owner
+    # Ensure new label is not one that already exists for that notebook
+    revisions = Revision.where(notebook_id: notebook.id)
+    revisions.each do |rev|
+      if rev.friendly_label == new_label
```
This throws an error if I accidentally click edit, and then immediately click save to not change the label.

mcrutch replied:
Ok, slightly different after I tried again. Click edit, change it, save it, click edit again, don't change it, click save, throws an error.

manfromjupyter replied:
Decided to do it through the front-end not via the backend so it doesn't even hit it (less times hitting our database).

mcrutch commented on app/controllers/revisions_controller.rb:
```diff
@@ -57,12 +57,43 @@ def latest_diff
+    errors = ""
+    notebook = Notebook.find(@revision.notebook_id)
+    friendly_label = params[:friendly_label].strip
+    if verify_revision_label(friendly_label, notebook, params[:old_label])
```
You are calling the same function twice with the same parameters.  Save the result the first time you call it and then append it to the errors. Don't call it twice.

manfromjupyter replied:
Thank you. Good catch. My mistake. Fixed.

mcrutch commented on app/controllers/revisions_controller.rb:
```diff
@@ -57,12 +57,43 @@ def latest_diff
     revision_summary = params[:summary].strip
     if revision_summary.length > 250
-      errors += "Revision summary was too long. Only accepts 250 characters and you submitted one that was #{revision_summary.length} characters."
+      errors += "Revision summary was too long. Only accepts a maximum of 250 characters and you submitted one that was #{revision_summary.length} characters. "
```
"Only accepts...." This still feels like clumsy grammar even with the edit
"Summary can be a maximum of 250 characters..." ?

manfromjupyter replied:
Fixed

mcrutch commented on app/controllers/change_requests_controller.rb:
```diff
@@ -154,9 +155,16 @@ def accept
+    # Revision Label Validation
+    if GalleryConfig.storage.track_revisions && params[:friendly_label] != ""
+      if verify_revision_label(params[:friendly_label], @notebook)
+        errors += verify_revision_label(params[:friendly_label], @notebook)
```
Another spot where you are double calling the same function.

manfromjupyter replied:
Thank you. Good catch. My mistake. Fixed.

manfromjupyter commented:
> Reviews table does not respect custom version numbers
> 
> * notebooks/#####/reviews
> * /reviews

Fixed this. Looked like the id to revision was only thing being used on review pages. Also fixed an issue where metadata Revision is technically the one being reviewed but that cannot receive a friendly_label.

mcrutch commented:
Found one more annoyance.  If you have a long revision number it is a bit awkward because the only way to get the pencil with mouse is to mouse over the version number itself and then move right because there is insufficient right padding to allow the glyphicon to be there otherwise
![image](https://user-images.githubusercontent.com/51446717/217351830-848b5930-3a9a-40c4-bce9-aebe1d3d47e5.png)

mcrutch commented on app/helpers/application_helper.rb:
```diff
@@ -79,10 +79,6 @@ def revision_path(rev)
     notebook_revision_path(rev.notebook, rev)
   end
 
-  def link_to_revision(rev)
```
For backwards compatibility, I think this should stay but at most throw a rails warning about deprecation.

manfromjupyter replied:
That makes sense. Will do

manfromjupyter commented:
> Found one more annoyance. If you have a long revision number it is a bit awkward because the only way to get the pencil with mouse is to mouse over the version number itself and then move right because there is insufficient right padding to allow the glyphicon to be there otherwise ![image](https://user-images.githubusercontent.com/51446717/217351830-848b5930-3a9a-40c4-bce9-aebe1d3d47e5.png)

I was worried about the column getting too wide but I see it's not as big of an issue as I thought. Fixed!

mcrutch commented on app/helpers/application_helper.rb:
```diff
@@ -80,6 +80,7 @@ def revision_path(rev)
   end
 
   def link_to_revision(rev)
+    logger.error('The rails helper "link_to_revision" is deprecated for this application. To ensure your revision links are able to show the user-friendly label, use the _link.slim partial instead.')
```
Not error, warn. and it must be called as Rails.logger.warn
`Rails.logger.warn("error message")`

manfromjupyter replied:
@mcrutch, done. Sorry, got mixed up with Django for the logger part.

mcrutch commented on app/views/application/_reviews.slim:
```diff
@@ -49,7 +49,7 @@
                   -if review.recent?
                     | Current
                   -else
-                    ==link_to_revision(review.revision)
```
is review.revision not already a revision object? (Same with below for row.revision)

manfromjupyter replied:
You're right. I had done this because I could not figure out why it was coming up as nothing because metadata and non-metadata revisions have the same commit_id. This has been fixed in the actual helper. Reverting this. Good catch.

mcrutch commented on app/views/revisions/index.slim:
```diff
@@ -81,7 +92,30 @@ div.content-container
+                    span.sr-only Edit revision label
+                  span.hidden aria-hidden="true" #{"]"}
+                form.revision-label-edit-form style="display: none" method="patch" data-url="#{revision_path(rev)}"
+                  -if rev.friendly_label != nil
```
Why do you need this differentiation. if rev.friendly_label is nil, it will just populate with empty values, no?  Just want to make sure we don't have an unnecessary if/then block that makes two almost identical lines sets of code for no reason.

manfromjupyter replied:
My bad. Yep. You're right. For some reason I thought a `nil` value in a string would cause an error.
